backends/source/dppl_sycl_queue_manager.cpp

```diff
@@ -0,0 +1,366 @@
+
+    switch (DeviceTy)
+    {
+    case DPPLSyclDeviceType::DPPL_CPU:
+    {
+        if (DNum >= cpu_queues.size()) {
+            std::stringstream ss;
+            ss << "SYCL CPU device " << DNum << " not found on system.";
+            error_reporter(ss.str());
+        }
+        active_queues[0] = cpu_queues[DNum];
+        break;
+    }
+    case DPPLSyclDeviceType::DPPL_GPU:
+    {
+        if (DNum >= gpu_queues.size()) {
+            std::stringstream ss;
+            ss << "SYCL GPU device " << DNum << " not found on system.";
+            error_reporter(ss.str());
+        }
+        active_queues[0] = gpu_queues[DNum];
+        break;
+    }
+    default:
+    {
+        error_reporter("Unsupported device type.");
+    }
+    }
+}
+
+/*!
+ * Allocates a new sycl::queue by copying from the cached {cpu|gpu}_queues
+ * vector. The pointer returned is now owned by the caller and must be properly
+ * cleaned up. The helper function DPPLDeleteSyclQueue() can be used is for that
+ * purpose.
+ */
+DPPLSyclQueueRef
+QMgrHelper::pushSyclQueue (DPPLSyclDeviceType DeviceTy, size_t DNum)
+{
+    queue *QRef = nullptr;
+    if(active_queues.empty())
+        error_reporter("Why is there no previous global context?");
+
+    switch (DeviceTy)
+    {
+    case DPPLSyclDeviceType::DPPL_CPU:
+    {
+        if (DNum >= cpu_queues.size()) {
+            std::stringstream ss;
+            ss << "SYCL CPU device " << DNum << " not found on system.";
+            error_reporter(ss.str());
+        }
+        active_queues.emplace_back(cpu_queues[DNum]);
+        QRef = new queue(active_queues[active_queues.size()-1]);
+        break;
+    }
+    case DPPLSyclDeviceType::DPPL_GPU:
+    {
+        if (DNum >= gpu_queues.size()) {
+            std::stringstream ss;
+            ss << "SYCL GPU device " << DNum << " not found on system.";
+            error_reporter(ss.str());
+        }
+        active_queues.emplace_back(gpu_queues[DNum]);
+        QRef = new queue(active_queues[active_queues.size()-1]);
+        break;
+    }
+    default:
+    {
+        error_reporter("Unsupported device type.");
+    }
+    }
+
+    return wrap(QRef);
+}
+
+/*!
+ * If there were any sycl::queue that were activated and added to the stack of
+ * activated queues then the top of the stack entry is popped. Note that since
+ * the same std::vector is used to keep track of the activated queues and the
+ * global queue a popSyclQueue call can never make the stack empty. Even
+ * after all activated queues are popped, the global queue is still available as
+ * the first element added to the stack.
+ */
+void
+QMgrHelper::popSyclQueue ()
+{
+    // The first queue which is the "default" queue can not be removed.
+    if(active_queues.size() <= 1 )
+        error_reporter("No active contexts");
+    active_queues.pop_back();
+}
+
+} /* end of anonymous namespace */
+
+/*!
+ * Returns the number of sycl::platform on the system.
+ */
+size_t DPPLGetNumPlatforms ()
```

Comment:
Not related to queue. I think it could be moved to separate file like `utilities` or `platform`.

Reply:
And duplicate in line 123.

Reply:
Moved the function to `dppl_sycl_platform_interface` in 7bb5faf